sql/stmtdiagnostics: TestTxnBundleCollection failed

sql/stmtdiagnostics.TestTxnBundleCollection [failed](https://mesolite.cluster.engflow.com/invocations/default/9b37efa7-6eed-4c19-b225-410c4a9f577d?testReportRun=1&testReportShard=3&testReportAttempt=1#targets-Ly9wa2cvc3FsL3N0bXRkaWFnbm9zdGljczpzdG10ZGlhZ25vc3RpY3NfdGVzdA==) on master @ [f9bdfb8b46ffa7d4e194cb82228f0d609df8e8ef](https://github.com/cockroachdb/cockroach/commits/f9bdfb8b46ffa7d4e194cb82228f0d609df8e8ef):


```
=== RUN   TestTxnBundleCollection
    test_log_scope.go:172: test logs captured to: outputs.zip/logTestTxnBundleCollection4071779001
    test_log_scope.go:82: use -show-logs to present logs inline
    statement_diagnostics_test.go:1149: -- test log scope end --
test logs left over in: outputs.zip/logTestTxnBundleCollection4071779001
--- FAIL: TestTxnBundleCollection (25.15s)
=== RUN   TestTxnBundleCollection/partial_match_transaction
    statement_diagnostics_test.go:1106: 
        	Error Trace:	pkg/sql/stmtdiagnostics/statement_diagnostics_test.go:1106
        	Error:      	Should be false
        	Test:       	TestTxnBundleCollection/partial_match_transaction
--- FAIL: TestTxnBundleCollection/partial_match_transaction (0.05s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=3</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnBundleCollection.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-63460

sql/stmtdiagnostics.TestTxnBundleCollection [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21333777?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21333777?buildTab=artifacts#/) on master @ [b072affffecee9534fb6b8a915babea09b1bb5d2](https://github.com/cockroachdb/cockroach/commits/b072affffecee9534fb6b8a915babea09b1bb5d2):


```
=== RUN   TestTxnBundleCollection
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/60db548121a3875fbc8a887cec8a69c0/logTestTxnBundleCollection65351105
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:168: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    statement_diagnostics_test.go:1149: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/60db548121a3875fbc8a887cec8a69c0/logTestTxnBundleCollection65351105
--- FAIL: TestTxnBundleCollection (24.64s)
=== RUN   TestTxnBundleCollection/partial_match_transaction
    statement_diagnostics_test.go:1106: 
        	Error Trace:	pkg/sql/stmtdiagnostics/statement_diagnostics_test.go:1106
        	Error:      	Should be false
        	Test:       	TestTxnBundleCollection/partial_match_transaction
--- FAIL: TestTxnBundleCollection/partial_match_transaction (0.04s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnBundleCollection.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

sql/stmtdiagnostics.TestTxnBundleCollection [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/21332707?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/21332707?buildTab=artifacts#/) on master @ [848e9d7823440526c0d2385b99923b6dacc31081](https://github.com/cockroachdb/cockroach/commits/848e9d7823440526c0d2385b99923b6dacc31081):


```
=== RUN   TestTxnBundleCollection
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/5344ad6650ee92bb69a16b66e4591733/logTestTxnBundleCollection1964626960
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:168: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    statement_diagnostics_test.go:1149: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/5344ad6650ee92bb69a16b66e4591733/logTestTxnBundleCollection1964626960
--- FAIL: TestTxnBundleCollection (28.87s)
=== RUN   TestTxnBundleCollection/partial_match_transaction
    statement_diagnostics_test.go:1106: 
        	Error Trace:	pkg/sql/stmtdiagnostics/statement_diagnostics_test.go:1106
        	Error:      	Should be false
        	Test:       	TestTxnBundleCollection/partial_match_transaction
--- FAIL: TestTxnBundleCollection/partial_match_transaction (0.03s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnBundleCollection.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

sql/stmtdiagnostics.TestTxnBundleCollection [failed](https://mesolite.cluster.engflow.com/invocations/default/b422292f-8a1d-40a9-a0c3-1004afce252d?testReportRun=5&testReportShard=3&testReportAttempt=1#targets-Ly9wa2cvc3FsL3N0bXRkaWFnbm9zdGljczpzdG10ZGlhZ25vc3RpY3NfdGVzdA==) on master @ [7cca024537d0efe15f607f4da37f401d198763ff](https://github.com/cockroachdb/cockroach/commits/7cca024537d0efe15f607f4da37f401d198763ff):


```
=== RUN   TestTxnBundleCollection
    test_log_scope.go:172: test logs captured to: outputs.zip/logTestTxnBundleCollection2378451336
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:168: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    statement_diagnostics_test.go:1149: -- test log scope end --
test logs left over in: outputs.zip/logTestTxnBundleCollection2378451336
--- FAIL: TestTxnBundleCollection (24.39s)
=== RUN   TestTxnBundleCollection/partial_match_transaction
    statement_diagnostics_test.go:1106: 
        	Error Trace:	pkg/sql/stmtdiagnostics/statement_diagnostics_test.go:1106
        	Error:      	Should be false
        	Test:       	TestTxnBundleCollection/partial_match_transaction
--- FAIL: TestTxnBundleCollection/partial_match_transaction (0.04s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=5</code>
 - <code>shard=3</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnBundleCollection.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

sql/stmtdiagnostics.TestTxnBundleCollection [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21334086?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21334086?buildTab=artifacts#/) on master @ [7cca024537d0efe15f607f4da37f401d198763ff](https://github.com/cockroachdb/cockroach/commits/7cca024537d0efe15f607f4da37f401d198763ff):


```
=== RUN   TestTxnBundleCollection
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/60db548121a3875fbc8a887cec8a69c0/logTestTxnBundleCollection3231392032
    test_log_scope.go:82: use -show-logs to present logs inline
    statement_diagnostics_test.go:1149: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/60db548121a3875fbc8a887cec8a69c0/logTestTxnBundleCollection3231392032
--- FAIL: TestTxnBundleCollection (24.15s)
=== RUN   TestTxnBundleCollection/partial_match_transaction
    statement_diagnostics_test.go:1106: 
        	Error Trace:	pkg/sql/stmtdiagnostics/statement_diagnostics_test.go:1106
        	Error:      	Should be false
        	Test:       	TestTxnBundleCollection/partial_match_transaction
--- FAIL: TestTxnBundleCollection/partial_match_transaction (0.06s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnBundleCollection.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

sql/stmtdiagnostics.TestTxnBundleCollection [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21334493?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21334493?buildTab=artifacts#/) on master @ [7c122f682c748419ef52d530daf7b4b7711570ba](https://github.com/cockroachdb/cockroach/commits/7c122f682c748419ef52d530daf7b4b7711570ba):


```
=== RUN   TestTxnBundleCollection
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/60db548121a3875fbc8a887cec8a69c0/logTestTxnBundleCollection1000551141
    test_log_scope.go:82: use -show-logs to present logs inline
    statement_diagnostics_test.go:1149: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/60db548121a3875fbc8a887cec8a69c0/logTestTxnBundleCollection1000551141
--- FAIL: TestTxnBundleCollection (22.47s)
=== RUN   TestTxnBundleCollection/partial_match_transaction
    statement_diagnostics_test.go:1106: 
        	Error Trace:	pkg/sql/stmtdiagnostics/statement_diagnostics_test.go:1106
        	Error:      	Should be false
        	Test:       	TestTxnBundleCollection/partial_match_transaction
--- FAIL: TestTxnBundleCollection/partial_match_transaction (0.06s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnBundleCollection.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

sql/stmtdiagnostics.TestTxnBundleCollection [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21334662?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21334662?buildTab=artifacts#/) on master @ [bca92f3b0e140f41ea2911e939d698e68eae8b06](https://github.com/cockroachdb/cockroach/commits/bca92f3b0e140f41ea2911e939d698e68eae8b06):


```
=== RUN   TestTxnBundleCollection
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/60db548121a3875fbc8a887cec8a69c0/logTestTxnBundleCollection1149483507
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:168: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    test_server_shim.go:637: DRPC is enabled (via testserver factory defaults)
    statement_diagnostics_test.go:1149: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/60db548121a3875fbc8a887cec8a69c0/logTestTxnBundleCollection1149483507
--- FAIL: TestTxnBundleCollection (22.25s)
=== RUN   TestTxnBundleCollection/partial_match_transaction
    statement_diagnostics_test.go:1106: 
        	Error Trace:	pkg/sql/stmtdiagnostics/statement_diagnostics_test.go:1106
        	Error:      	Should be false
        	Test:       	TestTxnBundleCollection/partial_match_transaction
--- FAIL: TestTxnBundleCollection/partial_match_transaction (0.08s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnBundleCollection.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/txn_instrumentation.go
+++ b/pkg/sql/txn_instrumentation.go
@@ -339,6 +339,17 @@ func (h *txnInstrumentationHelper) Finalize(
 		return
 	}
 	if collector.InProgress() {
+		// An InProgress collector at Finalize time corresponds to an implicit
+		// transaction: there is no terminal statement to drive the state
+		// machine through MaybeContinueDiagnostics. If any expected statement
+		// fingerprints remain uncaptured, the implicit transaction was a
+		// partial match for the request and must not be finalized.
+		if len(collector.stmtsFpsToCapture) > 0 {
+			log.Dev.VInfof(ctx, 2, "txn diag: aborting partial collection for "+
+				"request %d, txn %s, txn fingerprint %s, %d stmt fingerprints not captured",
+				collector.requestId, txnID, txnFpStr, len(collector.stmtsFpsToCapture))
+			return
+		}
 		log.Dev.VInfof(ctx, 2, "txn diag: finalizing collection for "+
 			"request %d, txn %s, txn fingerprint %s",
 			collector.requestId, txnID, txnFpStr)

